app/pubgrade_sidecar/controllers.py

```diff
@@ -0,0 +1,121 @@
+
+
+
+
+def getDeployment():
+    pod_list = apiV1.list_namespaced_deployment(namespace)
+    pods = []
+    for pod in pod_list.items:
+        pods.append(pod.metadata.name)
+    return pods
+
+
+def getImage(deployment_name: str):
+    access_token = current_app.config["FOCA"].environments["secrets"][
+        "access_token"
+    ]
+    x_access_token = request.headers["X-Access-Token"]
+
+    if access_token != x_access_token:
+        raise Unauthorized
+
+    try:
+        deployment = apiV1.read_namespaced_deployment(
+            deployment_name, namespace
+        )
+        return deployment.spec.template.spec.containers[0].image
+    except Exception:
+        raise NotFound
+
+
+def deleteDeployment(deployment_name: str):
+    access_token = current_app.config["FOCA"].environments["secrets"][
+        "access_token"
+    ]
+    x_access_token = request.headers["X-Access-Token"]
+
+    if access_token != x_access_token:
+        raise Unauthorized
+
+    try:
+        v1.delete_namespaced_pod(
+            deployment_name, namespace
+        )
+    except Exception:
+        raise NotFound
+    return "deleted: " + deployment_name
+
+
+def updateDeployment(deployment_name: str):
+    access_token = current_app.config["FOCA"].environments["secrets"][
+        "access_token"
+    ]
+    x_access_token = request.headers["X-Access-Token"]
+
+    if access_token != x_access_token:
+        raise Unauthorized
+
+    image_repo = request.json["image_name"]
+    image_tag = request.json["tag"]
+    old_image = getImage(deployment_name)
+    image = image_repo + ":" + image_tag
+    if re.split(":", old_image)[0] != re.split(":", image)[0]:
+        return "Cannot change image, only tag"
+
+    patch = [
+        {
+            "op": "replace",
+            "value": image,
+            "path": "/spec/template/spec/containers/0/image",
+        }
+    ]
+
+    # TODO: Re write retry logic with kafka pub-sub queue, separating retry from server process into a different pod
+    with ThreadPoolExecutor() as executor:
+        executor.submit(
+            update_image, image_repo, image_tag, deployment_name, patch, 0
+        )
+        return jsonify({"message": "Updating image in progress..."})
+
+
+def update_image(image_repo, image_tag, deployment_name, patch, retry):
+    if retry > MAX_RETRY:
+        raise DockerImageUnavailable
+    if check_docker_image_availability(image_repo, image_tag):
+        return apiV1.patch_namespaced_deployment(
+            name=deployment_name, namespace=namespace, body=patch
+        )
+    else:
+        time.sleep(RETRY_INTERVAL)
+        update_image(image_repo, image_tag, deployment_name, patch, retry + 1)
+
+
+def check_docker_image_availability(image_name, tag='latest'):
+    base_url = 'https://hub.docker.com'
+    url = f'{base_url}/v2/repositories/{image_name}/tags/{tag}/'
+    response = requests.get(url)
+    if response.status_code == 200:
+        return True
+    else:
+        return False
```

Comment:
nitpick, this should be:

```suggestion
    return response.status_code == 200
```